### PR TITLE
Datasource docs update, include boot status

### PR DESCRIPTION
**Pull request by holmanb**
## Additional Context
Blocked by https://github.com/canonical/cloud-init/pull/4578, separate PR because that one is already too big.

## Merge type

- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by holmanb** [comment]
@TheRealFalcon Commit 717cc98c4af46f9 is a [continuation of the conversation](https://github.com/canonical/cloud-init/pull/4578#discussion_r1416231087) from https://github.com/canonical/cloud-init/pull/4578, since boot status was otherwise undocumented. Would you mind reviewing this PR?

This PR includes a couple of other miscellaneous commits.

I also added some content for requirements for new datasources, including a section about reusing existing datasources (i.e. libvirt, proxmox), more logistical expectations for new datasources (access to testing, an informal maintainer), as well as a short section describing the benefits of upstreaming your datasource.

This includes a bump to the link count that linkcheck can't follow in Github Actions for some reason, as well as prints the list of broken links in the job log.

File: doc/rtd/development/datasource_creation.rst
Comment (TheRealFalcon) [inline doc/rtd/development/datasource_creation.rst:16]: Should we highlight `DatasourceNoCloud.py` or have it be a link?

**Review by TheRealFalcon — APPROVED** [review]
Nice add.

I left some inline comments, but none of them blocking. Take or leave what you want.

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:13]: This sounds a bit passive. I'd say "To add cloud-init support for a new platform, there are two possible approaches".

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:21]:
```suggestion
   option is more sensible for clouds that have a unique architecture.
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:26]:
```suggestion
There are some technical and logistical prerequisites that must be met for
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:29]:
```suggestion
Technical requirements
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:34]:
```suggestion
A cloud needs to be able to identify itself to cloud-init at
runtime, and be able to provide configuration to the
instance.
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:58]:
```suggestion
There are different methods that cloud-init can use to retrieve
cloud-configuration for configuring the instance. The most
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:61]:
```suggestion
Logistical requirements
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:72]:
```suggestion
A platform that isn't available for testing cannot be independently
validated. You will need to provide some means for community members and
upstream developers to test and verify this platform. Code that cannot be used
cannot be supported.
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:82]:
```suggestion
A point of contact is required who can answer questions and occasionally
provide testing or maintenance support. Maintainership is relatively informal,
but there is an expectation that from time to time upstream may
need to contact the maintainer with inquiries. Datasources that appear
to be unmaintained and/or unused may be considered for eventual
removal.
```

File: doc/rtd/development/datasource_creation.rst
Comment (s-makin) [inline doc/rtd/development/datasource_creation.rst:171]:
```suggestion
Datasources included in upstream cloud-init benefit from ongoing
```

File: doc/rtd/howto/status.rst
Comment (s-makin) [inline doc/rtd/howto/status.rst:64]:
```suggestion
Separately from the current running status described above, cloud-init can also
```

File: doc/rtd/howto/status.rst
Comment (s-makin) [inline doc/rtd/howto/status.rst:83]:
```suggestion
- ``'unknown'``: ``ds-identify`` has not run yet to determine if cloud-init should
  be run during this boot
- ``'disabled-by-marker-file'``: :file:`/etc/cloud/cloud-init.disabled` exists
  which prevents cloud-init from ever running
- ``'disabled-by-generator'``: ``ds-identify`` determined no applicable cloud-init
  datasources
- ``'disabled-by-kernel-cmdline'``: environment variable ``KERNEL_CMDLINE`` contained
  ``cloud-init=disabled``
- ``'disabled-by-environment-variable'``: environment variable ``KERNEL_CMDLINE``
  contained ``cloud-init=disabled``
- ``'enabled-by-kernel-cmdline'``: ``KERNEL_CMDLINE`` contained
  ``cloud-init=enabled``
- ``'enabled-by-generator'``: ``ds-identify`` detected possible cloud-init
  datasources
- ``'enabled-by-sysvinit'``: enabled by default in SysV init environment
```

**Review by s-makin — COMMENTED** [review]
Just some minor nits and wording suggestions, otherwise LGTM :)

File: doc/rtd/howto/status.rst
Comment (holmanb) [reply doc/rtd/howto/status.rst:83]: One exception here: "``KERNEL_CMDLINE`` contained ``cloud-init=enabled``" is not right. In this case we are talking literally about the kernel command line. The line above is referring to an environment variable with the name "KERNEL_CMDLINE". The kernel commandline itself can be read via `cat /proc/cmdline` and is different from the aforementioned environment variable with a similar name.

**Comment by holmanb** [comment]
Thanks for the reviews @TheRealFalcon and @s-makin. Great suggestions all around :-)

File: doc/rtd/howto/status.rst
Comment (s-makin) [reply doc/rtd/howto/status.rst]: Thanks for the context - I saw the discrepancy and figured they were all meant to be the same.

**Review by s-makin — APPROVED** [review]

File: doc/rtd/howto/status.rst
Comment (s-makin) [reply doc/rtd/howto/status.rst]: (to disambiguate, I would probably propose changing cmdline to command line, but I leave that to your discretion - happy to approve either way)

Comment (holmanb) [reply doc/rtd/howto/status.rst]: agreed, looks like we can drop commandline from the spelling list too

**Pull request merged**
